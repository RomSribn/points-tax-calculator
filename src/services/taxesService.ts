import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';
import { ITaxesResponse } from '@utils/interfaces';
import { apiUrl } from '@utils/config';
/**
 * Main API service.
 */
const taxesService = createApi({
  reducerPath: 'taxesApi',
  baseQuery: fetchBaseQuery({ baseUrl: apiUrl }),
  tagTypes: ['Taxes'],
  endpoints: (build) => ({
    getTaxesBrackets: build.query<ITaxesResponse, number>({
      query: (year = 2022) => ({
        url: `/tax-calculator/tax-year/${year}`
      }),
      providesTags: () => ['Taxes']
    })
  })
});

export { taxesService };
