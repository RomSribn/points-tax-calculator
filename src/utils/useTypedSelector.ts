import { TypedUseSelectorHook, useSelector } from 'react-redux';
import { RootState } from '@redux/store';
/**
 * Typed useSelector.
 */
export const useTypedSelector: TypedUseSelectorHook<RootState> = useSelector;
