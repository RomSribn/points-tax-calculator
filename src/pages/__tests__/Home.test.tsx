import React from 'react';
import { customRender, screen } from '@utils/tests';
import { Home } from '@pages/index';
import { useTypedSelector } from '@utils/useTypedSelector';
import { taxesService } from '@services/taxesService';

jest.mock('@utils/useTypedSelector');
jest.mock('@services/taxesService');

describe('Home component', () => {
  const appMockedState = {
    income: 100000,
    currentBracket: {
      max: 210371,
      min: 147667,
      rate: 0.29
    },
    year: 2019,
    result: 71000
  };

  const apiMockedState = { isLoading: true, isFetching: true };

  beforeEach(() => {
    (useTypedSelector as jest.Mock).mockReturnValue({});
    (taxesService.useGetTaxesBracketsQuery as jest.Mock).mockReturnValue({});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Should renders correctly', async () => {
    const wrapper = customRender(<Home />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Should renders TaxesForm and ResultCard when data is available', () => {
    // Mock the response data for useGetTaxesBracketsQuery
    (useTypedSelector as jest.Mock).mockReturnValue(appMockedState);

    const wrapper = customRender(<Home />);

    expect(screen.getByTestId('taxes-form')).toBeInTheDocument();
    expect(screen.getByTestId('result-card')).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });

  test('Should renders Spinner when loading data', () => {
    // Mock the loading state for useGetTaxesBracketsQuery
    (taxesService.useGetTaxesBracketsQuery as jest.Mock).mockReturnValue(apiMockedState);

    const wrapper = customRender(<Home />);

    expect(screen.getByTestId('taxes-form-spinner')).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });

  test('Should renders Spinner when fetching data', () => {
    // Mock the fetching state for useGetTaxesBracketsQuery
    (taxesService.useGetTaxesBracketsQuery as jest.Mock).mockReturnValue({ ...apiMockedState, isLoading: false });

    const wrapper = customRender(<Home />);

    expect(screen.getByTestId('result-spinner')).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
});
