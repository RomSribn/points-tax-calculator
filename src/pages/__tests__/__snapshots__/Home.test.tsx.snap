// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Should renders Spinner when fetching data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="vh-100 d-flex align-items-center"
        data-testid="home-page"
      >
        <div
          class="container"
        >
          <div
            class="row d-flex justify-content-center"
          >
            <div
              class="col-md-4"
            >
              <div
                class="mb-5"
              >
                <div
                  class="form-access my-auto"
                  data-testid="taxes-form"
                >
                  <form>
                    <span>
                      Tax calculate
                    </span>
                    <div
                      aria-live="polite"
                      class=""
                    >
                      <div
                        class="form-group"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="income"
                          class="form-control"
                          id="income"
                          name="income"
                          placeholder="Type your income - $"
                          type="text"
                        />
                      </div>
                    </div>
                    <div
                      aria-live="polite"
                      class=""
                    >
                      <div
                        class="form-group"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="year"
                          class="form-control"
                          id="year"
                          name="year"
                          placeholder="Type your tax year"
                          type="text"
                        />
                      </div>
                    </div>
                    <button
                      class="btn btn-primary"
                      disabled=""
                      type="submit"
                    >
                      Calculating...
                    </button>
                  </form>
                </div>
              </div>
            </div>
            <div
              class="col-md-6 d-flex align-items-center justify-content-center"
              data-testid="result-spinner"
            >
              <div
                class="d-flex align-items-center justify-content-center"
                data-testid="spinner-container"
              >
                <div
                  class="spinner-border text-primary"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="vh-100 d-flex align-items-center"
      data-testid="home-page"
    >
      <div
        class="container"
      >
        <div
          class="row d-flex justify-content-center"
        >
          <div
            class="col-md-4"
          >
            <div
              class="mb-5"
            >
              <div
                class="form-access my-auto"
                data-testid="taxes-form"
              >
                <form>
                  <span>
                    Tax calculate
                  </span>
                  <div
                    aria-live="polite"
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <input
                        aria-invalid="false"
                        aria-label="income"
                        class="form-control"
                        id="income"
                        name="income"
                        placeholder="Type your income - $"
                        type="text"
                      />
                    </div>
                  </div>
                  <div
                    aria-live="polite"
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <input
                        aria-invalid="false"
                        aria-label="year"
                        class="form-control"
                        id="year"
                        name="year"
                        placeholder="Type your tax year"
                        type="text"
                      />
                    </div>
                  </div>
                  <button
                    class="btn btn-primary"
                    disabled=""
                    type="submit"
                  >
                    Calculating...
                  </button>
                </form>
              </div>
            </div>
          </div>
          <div
            class="col-md-6 d-flex align-items-center justify-content-center"
            data-testid="result-spinner"
          >
            <div
              class="d-flex align-items-center justify-content-center"
              data-testid="spinner-container"
            >
              <div
                class="spinner-border text-primary"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Home component Should renders Spinner when loading data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="vh-100 d-flex align-items-center"
        data-testid="home-page"
      >
        <div
          class="container"
        >
          <div
            class="row d-flex justify-content-center"
          >
            <div
              class="col-md-4"
            >
              <div
                data-testid="taxes-form-spinner"
              >
                <div
                  class="d-flex align-items-center justify-content-center"
                  data-testid="spinner-container"
                >
                  <div
                    class="spinner-border text-primary"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="vh-100 d-flex align-items-center"
      data-testid="home-page"
    >
      <div
        class="container"
      >
        <div
          class="row d-flex justify-content-center"
        >
          <div
            class="col-md-4"
          >
            <div
              data-testid="taxes-form-spinner"
            >
              <div
                class="d-flex align-items-center justify-content-center"
                data-testid="spinner-container"
              >
                <div
                  class="spinner-border text-primary"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Home component Should renders TaxesForm and ResultCard when data is available 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="vh-100 d-flex align-items-center"
        data-testid="home-page"
      >
        <div
          class="container"
        >
          <div
            class="row d-flex justify-content-center"
          >
            <div
              class="col-md-4"
            >
              <div
                class="mb-5"
              >
                <div
                  class="form-access my-auto"
                  data-testid="taxes-form"
                >
                  <form>
                    <span>
                      Tax calculate
                    </span>
                    <div
                      aria-live="polite"
                      class=""
                    >
                      <div
                        class="form-group"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="income"
                          class="form-control"
                          id="income"
                          name="income"
                          placeholder="Type your income - $"
                          type="text"
                        />
                      </div>
                    </div>
                    <div
                      aria-live="polite"
                      class=""
                    >
                      <div
                        class="form-group"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="year"
                          class="form-control"
                          id="year"
                          name="year"
                          placeholder="Type your tax year"
                          type="text"
                        />
                      </div>
                    </div>
                    <button
                      class="btn btn-primary"
                      type="submit"
                    >
                      Calculate
                    </button>
                  </form>
                </div>
              </div>
            </div>
            <div
              class="col-md-6"
            >
              <div
                class="card"
                data-testid="result-card"
              >
                <div
                  class="card-body"
                >
                  <h5
                    class="card-title"
                  >
                    Result
                  </h5>
                  <div
                    class="settings-notification"
                  >
                    <ul>
                      <li>
                        <div
                          class="notification-info"
                        >
                          <p>
                            Tax rate
                          </p>
                          <span>
                            Total tax rate for the 
                            2019
                             year
                          </span>
                        </div>
                        <p>
                          28.999999999999996
                          %
                        </p>
                      </li>
                      <li>
                        <div
                          class="notification-info"
                        >
                          <p>
                            Total pay
                          </p>
                          <span>
                            Your gross salary income
                          </span>
                        </div>
                        <p>
                          100,000
                          $
                        </p>
                      </li>
                      <li>
                        <div
                          class="notification-info"
                        >
                          <p>
                            Taxes pay
                          </p>
                          <span>
                            Tax you pay
                          </span>
                        </div>
                        <p>
                          29,000
                          $
                        </p>
                      </li>
                      <li
                        class="text-success"
                      >
                        <div
                          class="notification-info"
                        >
                          <p>
                            Net pay
                          </p>
                          <span />
                        </div>
                        <p>
                          71,000
                          $
                        </p>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="vh-100 d-flex align-items-center"
      data-testid="home-page"
    >
      <div
        class="container"
      >
        <div
          class="row d-flex justify-content-center"
        >
          <div
            class="col-md-4"
          >
            <div
              class="mb-5"
            >
              <div
                class="form-access my-auto"
                data-testid="taxes-form"
              >
                <form>
                  <span>
                    Tax calculate
                  </span>
                  <div
                    aria-live="polite"
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <input
                        aria-invalid="false"
                        aria-label="income"
                        class="form-control"
                        id="income"
                        name="income"
                        placeholder="Type your income - $"
                        type="text"
                      />
                    </div>
                  </div>
                  <div
                    aria-live="polite"
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <input
                        aria-invalid="false"
                        aria-label="year"
                        class="form-control"
                        id="year"
                        name="year"
                        placeholder="Type your tax year"
                        type="text"
                      />
                    </div>
                  </div>
                  <button
                    class="btn btn-primary"
                    type="submit"
                  >
                    Calculate
                  </button>
                </form>
              </div>
            </div>
          </div>
          <div
            class="col-md-6"
          >
            <div
              class="card"
              data-testid="result-card"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Result
                </h5>
                <div
                  class="settings-notification"
                >
                  <ul>
                    <li>
                      <div
                        class="notification-info"
                      >
                        <p>
                          Tax rate
                        </p>
                        <span>
                          Total tax rate for the 
                          2019
                           year
                        </span>
                      </div>
                      <p>
                        28.999999999999996
                        %
                      </p>
                    </li>
                    <li>
                      <div
                        class="notification-info"
                      >
                        <p>
                          Total pay
                        </p>
                        <span>
                          Your gross salary income
                        </span>
                      </div>
                      <p>
                        100,000
                        $
                      </p>
                    </li>
                    <li>
                      <div
                        class="notification-info"
                      >
                        <p>
                          Taxes pay
                        </p>
                        <span>
                          Tax you pay
                        </span>
                      </div>
                      <p>
                        29,000
                        $
                      </p>
                    </li>
                    <li
                      class="text-success"
                    >
                      <div
                        class="notification-info"
                      >
                        <p>
                          Net pay
                        </p>
                        <span />
                      </div>
                      <p>
                        71,000
                        $
                      </p>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Home component Should renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="vh-100 d-flex align-items-center"
        data-testid="home-page"
      >
        <div
          class="container"
        >
          <div
            class="row d-flex justify-content-center"
          >
            <div
              class="col-md-4"
            >
              <div
                class="mb-5"
              >
                <div
                  class="form-access my-auto"
                  data-testid="taxes-form"
                >
                  <form>
                    <span>
                      Tax calculate
                    </span>
                    <div
                      aria-live="polite"
                      class=""
                    >
                      <div
                        class="form-group"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="income"
                          class="form-control"
                          id="income"
                          name="income"
                          placeholder="Type your income - $"
                          type="text"
                        />
                      </div>
                    </div>
                    <div
                      aria-live="polite"
                      class=""
                    >
                      <div
                        class="form-group"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="year"
                          class="form-control"
                          id="year"
                          name="year"
                          placeholder="Type your tax year"
                          type="text"
                        />
                      </div>
                    </div>
                    <button
                      class="btn btn-primary"
                      type="submit"
                    >
                      Calculate
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="vh-100 d-flex align-items-center"
      data-testid="home-page"
    >
      <div
        class="container"
      >
        <div
          class="row d-flex justify-content-center"
        >
          <div
            class="col-md-4"
          >
            <div
              class="mb-5"
            >
              <div
                class="form-access my-auto"
                data-testid="taxes-form"
              >
                <form>
                  <span>
                    Tax calculate
                  </span>
                  <div
                    aria-live="polite"
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <input
                        aria-invalid="false"
                        aria-label="income"
                        class="form-control"
                        id="income"
                        name="income"
                        placeholder="Type your income - $"
                        type="text"
                      />
                    </div>
                  </div>
                  <div
                    aria-live="polite"
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <input
                        aria-invalid="false"
                        aria-label="year"
                        class="form-control"
                        id="year"
                        name="year"
                        placeholder="Type your tax year"
                        type="text"
                      />
                    </div>
                  </div>
                  <button
                    class="btn btn-primary"
                    type="submit"
                  >
                    Calculate
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
